### DISCUSSION TITLE
how to make dynamic db url?

### DISCUSSION THREAD

**Discussion by yeukfei02** [prisma/prisma#3561] [Q&A] [answered]
```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model users {
  users_id   Int      @default(autoincrement()) @id
  email      String
  password   String
  created_by DateTime @default(now())
  updated_by DateTime @default(now())
}
```

this is `schema.prisma`, how to make dynamic URL in this file?
or in node how do you make it?

I want to do  
`process.env.NODE_ENV === 'development'` > connect localhost db
else connect clould/docker db

thanks.

**Comment by ryands17** [comment]
@yeukfei02 👋 
You can use [dotenv-cli](https://www.npmjs.com/package/dotenv-cli) to load different environment files for development and production.

For e.g. in development

```
dotenv -e .env.development -- node index.js
```

This will load the `.env.development` file. And you can do the same in production.

**Reply by TheoMer** [reply]
@ryands17 So I moved the .env file out of the prisma folder into root, set dev on my package.json to:

```
dotenv -e .env -- ts-node-dev --inspect --no-notify --respawn --transpileOnly src/index
```

Annnnd, it doesn't work. I also tried setting `url = process.env.DATABASE_URL` and that didn't work either. What am I overlooking?

**Reply by ryands17** [reply]
Any error showing up? If so could you share a reproduction as well if possible?

**Reply by TheoMer** [reply]
@ryands17 I get the following error message when the 'me' query (the main gateway into the site):

```ts
import { gql, useQuery } from '@apollo/client';
import PropTypes from 'prop-types';

const CURRENT_USER_QUERY = gql`
  query {
    me {
      id
      email
      name
      password
      permissions2
      address {
        id
        card_name
        address_line
        city
        postcode
        country
      }
      order {
        id
      }
      items {
        id
        price
        image
        largeImage
        image2
        largeImage2
        image3
        largeImage3
        image4
        largeImage4
        image5
        largeImage5
        image6
        largeImage6
        title
        description
        mainDescription
        quantity
        Color {
          id
          name
          label
        }
        Size {
          id
          name
          label
        }
        itemvariants {
          id
          price
          image
          largeImage
          image2
          largeImage2
          image3
          largeImage3
          image4
          largeImage4
          image5
          largeImage5
          image6
          largeImage6
          title
          description
          mainDescription
          quantity
          Color {
            id
            name
            label
          }
          Size {
            id
            name
            label
          }
          item
        }
      }
      cart {
        id
        quantity
        ItemVariants {
          id
          price
          image
          largeImage
          title
          description
          mainDescription
          quantity
          Color {
            id
            name
            label
          }
          Size {
            id
            name
            label
          }
          item
        }
        Item {
          id
          price
          image
          largeImage
          title
          description
          mainDescription
          quantity
          Color {
            id
            name
            label
          }
          Size {
            id
            name
            label
          }
          itemvariants {
            id
            price
            image
            largeImage
            title
            description
            mainDescription
            quantity
            Color {
              id
              name
              label
            }
            Size {
              id
              name
              label
            }
            item
          }
        }
      }
    }
  }
`;

// User
function useUser() {

  const { ...result } = useQuery(CURRENT_USER_QUERY);

  if (result.data) {
    return result;
  }
  
}

export { CURRENT_USER_QUERY, useUser };


```

is actioned:

```
Invalid `ctx.prisma.user.create()` invocation in
C:\Users\d0475\Documents\Projects\flamingo-ecom2\sick-fits\backend\src\types\resolvers\Query.ts:90:52

   86 const password = await bcrypt.hash(newUUID, 10);
   87 const name = "Guest User";
   88
   89 // 3. create the user in the database
→  90 const user = await ctx.prisma.user.create(
  Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: TableDoesNotExist { table: "flamingo-production.User" } })
me err = TypeError: Cannot read property 'id' of undefined
```

As can be seen the User table does exist:

![User_table](https://user-images.githubusercontent.com/15969262/92747482-25916000-f37c-11ea-8dab-de0c0482a0f0.jpg)


You can find my latest repo here: https://github.com/TheoMer/Prisma2-Miration

**Reply by ryands17** [reply]
This doesn't seem to me like a `dotenv` or env issue. Prisma can find the env and connect to the DB but the table doesn't exist. Could you verify the DB url and run `prisma generate` again and check?

**Reply by TheoMer** [reply]
Yep, all is in order with the dB url and the existence of the table. If I place .env back in the prisma folder everything works perfectly as expected.

**Reply by ryands17** [reply]
It's working for me. I tried with the same env file in your repo (`variables2.env`) and it works fine. The only thing I did was install `dotenv-cli` locally in the repo as I didn't have that globally.

Also your `dev` command was giving issue so I had to replace it to this

```json
"dev": "dotenv -e variables2.env -- ts-node-dev --no-notify --respawn --transpile-only src/index"
```

**Reply by TheoMer** [reply, marked as answer]
Hey @ryands17 Thanks for the heads up on the dev script. I finally got it working by refactoring my index file to:

```
import { config } from 'dotenv';
const path = process.env.NODE_ENV === 'production' ? 'variables.production.env' : 'variables.env';
config({ path });
```

**Reply by yeukfei02** [reply]
@TheoMer
 where to add this?
`variables.production.env` u put into prisma folder or main folder?

**Reply by TheoMer** [reply]
@ryands17 This I have situated in the root folder.

**Comment by rocketteering** [comment]
There is a better way, which I found out after few hours of digging for answer:

When you initiate Prisma Client, you can pass the URL-dynamically and use whatever variables you like.

In my case:

```
const prisma = new PrismaClient({
    datasources: {
      db: {
        url: `mysql://${DB_USERNAME}:${DB_PASSWORD}@${DB_HOST}/${DB_NAME}`,
      },
    },
  })
```

Then change url in `prisma.schema` file to following, since it can't be empty

```
datasource db {
  provider = "mysql"
  url = ""
}
```

And remember from [documentation](https://www.prisma.io/docs/reference/database-reference/connection-urls:)

> Special characters
> For MySQL and PostgreSQL, you must percentage-encode special characters in any part of your connection URL - including passwords. For example, p@$$w0rd becomes p%40%24%24w0rd.

**Reply by kennyma603** [reply]
i am using this approach, however, after i made a schema change, and do `npx prisma migrate dev`, it throws error saying the connection url is empty.

Prisma schema loaded from prisma\schema.prisma
Datasource "db"

Error: Connection url is empty.
@IlmariKu

**Reply by janpio** [reply]
Replied here: https://github.com/prisma/prisma/discussions/12336#discussioncomment-2361283

**Reply by rocketteering** [reply]
I think you can just use a dummy-value in place of url like @janpio mentioned.

**Reply by royzhang107** [reply]
Using the .env file and expanding the environment variables should solve the problem.

` DATABASE_URL=mysql://${DB_USERNAME}:${DB_PASSWORD}@${DB_HOST}/${DB_NAME}`

refs: [docs](https://www.prisma.io/docs/guides/development-environment/environment-variables#expanding-variables)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
